### ISSUE TITLE
Positive transform fails large numbers

### ISSUE THREAD

**Issue by mrksr** [GPflow/GPflow#302]
The positive transform implemented in `GPflow.transforms.positive` is an alias for the `Log1pe`-transform. The name suggests that it can be used to represent arbitrary positive numbers, however, it is actually only possible to to represent numbers up a value of about 700 due to the result of `np.exp` growing too large:

```python
In [31]: GPflow.transforms.positive.forward(700)
Out[31]: 700.000001

In [32]: GPflow.transforms.positive.forward(800)
/(...)/GPflow/transforms.py:129: RuntimeWarning: overflow encountered in exp
  return np.log(1. + np.exp(x)) + self._lower
Out[32]: inf

In [33]: GPflow.transforms.positive.backward(700)
Out[33]: array([ 699.999999])

In [34]: GPflow.transforms.positive.backward(800)
/(...)/GPflow/transforms.py:138: RuntimeWarning: overflow encountered in exp
  return np.log(np.exp(y - self._lower) - np.ones(1, np_float_type))
Out[34]: array([ inf])
```

Since the positive transform degrades to an identity function for large values, it should be fairly easy to fix this. However, it is probably important to ensure that both `forward` and `tf_forward` behave the same.

I guess a similar problem will exist for the `Logistic`-transform.

**Comment by markvdw** [comment]
This is also highlighted in https://github.com/GPflow/GPflow/issues/300. I didn't think this was all too much of a problem, since values of 700 would be odd to encounter, but given the simplicity of the fix, it may be worth doing.

**Comment by jameshensman** [comment]
We should fix this, marking as a bug.

If the optimal value is out of range, then we'll *not* get an overflow during optimization, because the tensorflow representation of the forward transform won't overflow. But when the optimization finishes, we get the overflow when putting the optimized vector back into the model state. Then, a successive call to optimize will fail, because the reverse transform (to initialize the optimizer) will fail.

**Comment by alexggmatthews** [comment]
tf.nn.softplus should do the trick:

https://www.tensorflow.org/api_docs/python/nn/activation_functions_#softplus

I'll have a look for a fix for logistic later I have a meeting with Zoubin this morning.

**Comment by jameshensman** [comment]
We already use tf.nn.softplus on the tensorflow side, the problem comes when setting the values, at the end of optimization, which uses the numpy transform. I'll push a fix for Log1pe, which I think is important. I'll leave Logistic for you, @alexggmatthews :)

**Comment by alexggmatthews** [comment]
OK sorry for the misunderstanding.

**Comment by mrksr** [comment]
I have played around a bit with the `Logistic` transform and I think it can be left as-is actually, since the only exponentiation happens in a forward-transform and within floating point accuracy, it is not really possible to get values through the backward-transform which might be a problem in forward direction (if that makes sense):

```python
In []: lg = GPflow.transforms.Logistic(0, 1)

In []: close_to_1 = 0.9999999999999999

In []: actually_1 = 0.99999999999999999

In []: lg.backward(close_to_1)
Out[]: 36.043653389117154

In []: lg.forward(lg.backward(close_to_1))
Out[]: 0.99999999999999978

In []: lg.backward(actually_1)
/nosave/sava_0/kaiser/python/GPflow/GPflow/transforms.py:159: RuntimeWarning: divide by zero encountered in log
  return -np.log((self.b - self.a) / (y - self.a) - 1.)
Out[]: inf

In []: close_to_1 == 1.
Out[]: False

In []: actually_1 == 1.
Out[]: True
```

It looks quite the same arount zero. I'd say while for the `Log1pe`-transform it was actually plausible to run into problems before the fix in 2049274 (thanks for that!), for the `Logistic` transform, the optimization itself would have to do fairly strange things.

**Comment by jameshensman** [comment]
Great investigating, @mrksr . 

I take it this is fixed by #303 , marking as closed. Happy Christmas!
